### ISSUE TITLE
Parse error in Haskell modules when module header is after LANGUAGE pragmas

### ISSUE THREAD

**Issue by chshersh** [vaclavsvejcar/headroom#28]
When a Haskell module contains LANGUAGE pragmas, the header is added at the top, the current header is not replaced. And this results in parsing error:

```
src/Summoner/Tree.hs:14:1: error:
    parse error on input ‘{- |
Copyright: (c) 2017-2019 Kowainik
SPDX-License-Identifier: MPL-2.0
Maintainer: Kowainik <xrom.xkov@gmail.com>

Data type for representing filesystem structure via tree.
-}’
   |
14 | {- |
   | ^^^^...
cabal: Failed to build documentation for summoner-1.4.0.0 (which is required
by summoner-tui-1.4.0.0).
```

The content of the file:

```haskell
{-|
Module      : MODULE_NAME
Description : SHORT_DESC
Copyright   : (c) 2020 Kowainik
License     : MIT
Maintainer  : xrom.xkov@gmail.com
Stability   : experimental
Portability : POSIX

LONG_DESC
-}
{-# LANGUAGE ViewPatterns #-}

{- |
Copyright: (c) 2017-2019 Kowainik
SPDX-License-Identifier: MPL-2.0
Maintainer: Kowainik <xrom.xkov@gmail.com>

Data type for representing filesystem structure via tree.
-}
```

**Comment by vaclavsvejcar** [comment]
@chshersh Thanks for reporting this, this (related to #25) is definitely something I'd like to focus on in next major release of _Headroom_. I'm just thinking how to process headers that are not at the top of the file (like this one). Should the newly rendered header replace the old one in-place, or remove the old one and add the new one at the very beginning? My concern about the first approach is that for files where the header already exists, it would be replaced in place (and put below the language pragmas in this case), but for the new files where's no header yet, the header would be placed at the very beginning, causing inconsistency across the codebase. What's your opinion about this?

**Comment by chshersh** [comment]
@chshersh I think the newly generated header should replace the existing one. In my projects I don't have a lot of modules. So, if by default `headroom` adds headers at the beginning of the module, I can still move newly added headers manually, and then consequent calls to `headroom` will update headers accordingly.

In my projects I'm following the style where top-level Haddock comment is placed after `{-# LANGUAGES #-}` and `{-# OPTIONS_GHC #-}` pragmas, so adding adding default comment after these lines would be more convenient for me 🙂

**Comment by mheinzel** [comment]
I ran into a related problem where some formatters (in this case [ormolu](https://github.com/tweag/ormolu])) always put the language extensions at the top. So headroom and ormolu keep fighting against each other and every time you run headroom after ormolu, it will add another header to the top.

My preference would be to add the header after language extensions. Probably the same for [other global options like `OPTIONS_GHC`](https://github.com/tweag/ormolu/blob/0233f6ddecc2521d5774d95765dbf7c618c60093/data/examples/other/pragma-out.hs) as well as [shebangs](https://github.com/tweag/ormolu/blob/master/data/examples/module-header/shebang-out.hs).

Not sure how compatible this is with other people's tools and preferences. Maybe there could be some configuration for it?

A way of handling similar issues in another tool can be seen here: https://github.com/cloudposse/copyright-header/blob/master/contrib/syntax.yml#L158

Let me know if there is anything I can help with to solve this issue.

**Comment by vaclavsvejcar** [comment]
@mheinzel Thank you for reporting this, I'm aware of this issue and unfortunately in current version of _Headroom_ there's no workaround for this. However, right now I'm working on major rewrite of internals (branch `v0200_rewrite`) that would allow me not only to fix this issue, but also the #13, #25, #28 and #31. Good point about the external configuration, I'll consider this as option to override the default behaviour in some external configuration (e.g. the configuration _YAML_ file) would be really nice to have. I'll let you know if I'll need any other details, thanks for your offer, and I'll also post updates here.

**Comment by vaclavsvejcar** [comment]
@chshersh @mheinzel I've spent last few days rewritting major part of the Headroom internals to make the header detection and manipulation logic more powerful. Notable changes are:
- Configuration is now composed using the _Partial Options Monoid_ pattern with default values stored in [/embedded/default-config.yaml](https://github.com/vaclavsvejcar/headroom/blob/master/embedded/default-config.yaml) (big thanks to @chshersh for suggesting this idea)
- Header detection patterns & position are now fully configurable in the `.headroom.yaml` file (thans @mheinzel for proposing this), you can basically freely define the header position by specifying before and after which pattern it should be present (`putBefore` & `putAfter`) and its starting and ending patterns (`startsWith` & `endsWith`). Again see the [/embedded/default-config.yaml](https://github.com/vaclavsvejcar/headroom/blob/master/embedded/default-config.yaml) for default configuration & feel free to modify it to suit your needs in `.headroom.yaml`.

Speaking about issues you both have with current Headroom placing the header at the very beginning instead of putting in after language pragmas, this new additional configuration added to `.headroom.yaml` should fix that:

```yaml
license-headers:
  haskell:
    put-after: ["^{-#"]
```

These changes are not yet released (will be part of the `0.2.0.0` release), but you can try them by checking out the `master` branch and building Headroom from it. If you'll have some spare time, please try this version with your projects and let me know if these changes matches your needs & solves your troubles with Headroom related to this issue. Please note that since this is not final release, you still may face some other bugs, also the _Haddock_ documentation is mostly missing.

**Comment by vaclavsvejcar** [comment]
Just realized that by requiring to specify the `starts-with` and`ends-with` I made impossible to use headers created with multiple single-line comments, instead of one multi-line one. This is solvable in languages that support both, but for example in Bash with only single-line comments, this would be issue. I'll think how to specify this nicely in the _YAML_ config and definitely implement that before releasing `0.2.0.0`, if you'd have any ideas regarding this, it would be highly appreciated.

**Comment by vaclavsvejcar** [comment]
Just added support for license headers that uses single-line comments. This is nice-to-have in languages that also support block comment, but must have to implement support for example for Bash (#35), which lacks block comment syntax. When specifying single line comment, one must use the `prefixed-by` property, instead of `starts-with` and `ends-with`, like this:

#### Haskell configuration (block comment header)
```yaml
license-headers:
  haskell:
    starts-with: "{-"
    ends-with: "-}"
```

#### Haskell configuration (single-line comment style header)
```yaml
license-headers:
  haskell:
    prefixed-by: "--"
```

Mixing `starts-with` and/or `ends-with` with `prefixed-by` is not allowed and will yield error when parsing YAML configuration file.

I still need to do proper documentation and better testing of this logic so there might be still some bugs, but it should be nice foundation for adding support for next languages and also to configure Headroom to specific needs of each user.

**Comment by mheinzel** [comment]
Looks like a nice solution.

Maybe it could be more explicit about being a "line" or "block" comment, either as part of the fields you proposed above or separate, but that's a matter of taste.

Also, for block comments, you might want to support an optional third configuration value (or re-use the "prefixed-by" one) to allow C-like block comments with a per-line "*" prefix.
It's quite a common style, see e.g. https://doc.rust-lang.org/rust-by-example/hello/comment.html and https://github.com/openalpr/openalpr/blob/master/src/main.cpp.

**Comment by vaclavsvejcar** [comment]
@mheinzel Thanks for sharing your thoughts.

About being more explicit - I was actually thinking about it, but haven't decided any nice way that wouldn't bring too much noise to configuration. Maybe something like this would work?

```yaml
license-header:
  haskell:
    file-extensions: ["hs"]
    block-comment:
      starts-with: "{-"
      ends-with: "-}"
```
or

```yaml
license-header:
  haskell:
    file-extensions: ["hs"]
    line-comment:
      prefixed-by: "--"
```

But not sure whether that isn't too much noisy.

Regarding the third optional value - not sure if you mean the same thing but you shouldn't need this with Headroom. The `starts-with` and `ends-with` are here just to tell Headroom how is the block comment delimited so it can detect it in source code file, but not how the rendered header will look like (that's what the content of the `*.mustache` template is for). So you could have

```yaml
starts-with: "/*"
ends-with: "*/"
```
in the configuration and still, this template:

```mustache
/*
 * Copyright (c) {{ year }} {{ author }}
 */
```

will be rendered like this

```c
/*
 * Copyright (c) 2020 John Smith
 */
```

as expected and also Headroom will detect such header just fine, as it starts with `/*` and ends with `*/`.

**Comment by mheinzel** [comment]
> But not sure whether that isn't too much noisy.

Looks good to me. Not sure if this will ever come up, but it also leaves open the option for the future of having both `block-comment` and `line-comment` configured (so both could be recognized), maybe with a preference for one, so it even rewrites one style into the other. Maybe one day someone wants that. :smile: 

> Regarding the third optional value

Ah, thanks. That makes sense.

I can think of one feature where this would still come in handy, but not sure if you want to support it:
If headroom knows how to prefix the individual lines, it can break the lines itself according to some maximum length in the configuration. This could for example be useful together with variables that get replaced, since in the template a user wouldn't know how long they will be (and thus, can't split lines correctly herself).

But yeah, those things are probably not necessary for most people and when you want to support them later, you can still change things.

**Comment by vaclavsvejcar** [comment]
>  ... it also leaves open the option for the future of having both `block-comment` and `line-comment` configured ...

Good point, I'll definitely consider that as future feature. For now, I'll focus to release the `0.2.0.0` with these core features on which the next releases can build. Anyway, in last commit I changed the structure of _YAML_ as discussed above, so for now, you need to define either `block-comment` (with `starts-with` and `ends-with` fields) or `line-comment` (with `prefixed-by` field).

Hopefully all these configuration options in YAML file will give end users enough power to adjust header position/appearance to their needs.

> I can think of one feature where this would still come in handy, but not sure if you want to support it

Also nice idea, thanks, I'll consider that for future releases 👍 

@mheinzel @chshersh In case that you'll retest the current `master` version, please let me know if that solves the problems you reported related to this issue, so we can eventually close it.

**Comment by mheinzel** [comment]
I gave it a try and while it mostly seems to work, there were some issues. Please have a look at the last 3 commits here: https://github.com/wireapp/wire-server/commits/licenses-headroom-0.2

1. First of all, let me know if the changes to `.headroom.yaml` were not correct.

2. For [files without language extensions](https://github.com/wireapp/wire-server/commit/fe919cdd9299183ea5de7a4fbb9822307651ae92#diff-0174e7d48bd986c18cb3e00271337287), the header gets put *below* the module name.

3. For [files with language extensions](https://github.com/wireapp/wire-server/commit/fe919cdd9299183ea5de7a4fbb9822307651ae92#diff-6a1beeb77588dbca6bf49f8019376191), it works, but puts a newline at the end. I'd rather be able to make it put one between language extension and license header, but that's a smaller issue.

4. After an initial run, [it adds additional newlines](https://github.com/wireapp/wire-server/commit/d8aa70391972e1b4ac942f580f66f3631a91e71f#diff-0174e7d48bd986c18cb3e00271337287), so it's not idempotent. (To find issues like this, it could be nice to have an optional check for idempotency that will make headroom run twice on each file and complain if the file changed the second time)

Issues like 3. and 4. are annoying in combination with our formatter, which wants this combination of language extension, comment and module header to be formatted like this:

```
{-# LANGUAGE OverloadedStrings #-}

-- some
-- license

module Foo where
```

but running headroom again turns it into

```
{-# LANGUAGE OverloadedStrings #-}
-- some
-- license



module Foo where
```

Would it be helpful for you if I would create some minimal files plus expected result for our usecase at Wire? Then it's easy to run headroom against it and we could see which changes (and maybe additional options) would work for this.

**Comment by vaclavsvejcar** [comment]
@mheinzel Thanks for trying that out and reporting the problems, I'll check it. I also plan to add some test suites (#37) to prevent these issues in the final `0.2.0.0` release. Will keep you updated.

**Comment by vaclavsvejcar** [comment]
@mheinzel I spent last few days by fixing some stuff & adding few (hopefully) useful new featues, here's the summary:

> 1. First of all, let me know if the changes to .headroom.yaml were not correct.

Your `.headroom.yaml` looks fine. I added few more options you might like to add - check below.

> 2. For files without language extensions, the header gets put below the module name.

Yep this was bug in Headroom, should be fixed.

> 3. and 4.

This was caused by the extra newline you have at the end of your `haskell.mustache` template. I was thinking about this issue and kinda reworked the internals like this: Now, any excessive newlines are stripped from the template, to prevent this "crawling newline" issue. If you want to specify that you'd like to have extra newline(s) before/after rendered template (as you mentioned in `3.`), you can now use new config keys, `margin-after` and `margin-before` with number of newlines to put before/after (default is 0 for both). In your case, you may want to have your configuration like this:

```yaml
license-headers:
  haskell:
    put-after: ["^{-#"]
    margin-before: 1
    margin-after: 1
    line-comment:
      prefixed-by: "--"
```

Also note that the `margin-before` newlines are not rendered when the comment is put at the very beginning of file (not sure if this is good idea or not yet)

---

These fixes should hopefully make Headroom usable for your needs. I quickly tested that with your `wire-served` project and looks fine at first sight (but please do some more careful tests). One more thing I noticed - some of your files starts like this:

```haskell
{-# OPTIONS_GHC -Wno-orphans -Wno-incomplete-patterns #-}

-- it's ok to not warn about unmatched patterns; 'validateEveryToJSON' will crash on them
-- before too long.

-- | swagger2 docs for galley generated with servant-swagger.  for now, this module contains
-- all of the servant code as well.
module Galley.API.Swagger
  ( GalleyRoutes,
    swagger,
  )
where
```

With your current settings, Headroom will consider the very first comment

```haskell
-- it's ok to not warn about unmatched patterns; 'validateEveryToJSON' will crash on them
-- before too long.
```

as valid header, because it matches the conditions (is put after `^{-#` and before `^module`, starts with `--`) and such comment will be either replaced (`run-mode: replace`) or skipped (`run-mode: add`). Not sure if I can do anything about that on Headroom side, as Headroom has no clue this is not license header, even if that matches all conditions.

Also if you'll test this again, I'd recommend trying that not from the current `HEAD` of your [licenses-headroom-0.2](https://github.com/wireapp/wire-server/compare/licenses-headroom-0.2) branch, but rather start from the commit [
20f8b1d](https://github.com/wireapp/wire-server/commit/20f8b1d3c0054928cf2dd3f22c8747e28a9e56b0), as now fixed version of Headroom might not be able to drop/replace some comment incorrectly added as described in `2.`.

**Comment by mheinzel** [comment]
>  the `margin-before` newlines are not rendered when the comment is put at the very beginning of file (not sure if this is good idea or not yet)

Seems like a good idea to me :+1: 

> One more thing I noticed

Yes, Headroom eats some of our comments, but I don't think it's big issue. It's only in a handful of places and I think they can be tweaked slightly to achieve a satisfactory result.

>  I'd recommend trying that not from the current HEAD of your licenses-headroom-0.2 branch

Makes sense, thanks! I'll give it another go.

**Comment by mheinzel** [comment]
Okay, I tested this for Haskell files now (Rust will have to wait until Monday).

There were only some files I needed to prepare (see [licenses-headroom-0.2](https://github.com/wireapp/wire-server/commits/licenses-headroom-0.2)), after that running Headroom inserted license headers in the correct places and the result was idempotent and also made Ormolu happy. :tada:

Even better:
On [licenses-headroom-onto-licensure](https://github.com/wireapp/wire-server/commits/licenses-headroom-onto-licensure), after making small changes to our current repo that already has license headers, running headroom is a no-op, so it correctly recognizes all license headers. And I guess changing the templace would replace the headers correctly.

*Some random notes*:

When preparing files, there was one interesting case:

```
{-# OPTIONS_GHC -Wno-missing-methods
    -Wno-orphans #-}
```

This caused the header to be passed as an argument to GHC. :D
It could be prevented by fiddling with the configuration, but for me seems okay to require pragmas to be single-line.

Also, for consistency with `--replace-headers` and `--drop-headers`, it might make sense to add  `--add-headers`.

**Comment by vaclavsvejcar** [comment]
@mheinzel Glad to hear it works for you and thank you for testing this 👍 There might still be some issues, I also plan to do some extensive testing before releasing the final `0.2.0.0` as I don't want to have public release that breaks code.

> When preparing files, there was one interesting case

Well, this is gonna be the tough one to solve... Currently, the `put-after`/`put-before` regular expressions are matched line-by-line, so Headroom just sees that the line `{-# OPTIONS_GHC -Wno-missing-methods` matches with the `^{-#` condition and just blindly puts the rendered header after this line. I'm not regexp expert, but what about to use such regular expression in `put-after` that would:

- match if single-line pragma used - `{-# OPTIONS_GHC -Wno-missing-methods -Wno-orphans  #-}`
- won't match if the pragma is not closed on current line (we expect that it will be closed on one of next lines)
- match if there's no opening part of the pragma on current line, but has closing part (we assue that this is the last line of the multi-line pragma)

In short, maybe something like this: `^({-#.*-}$)|(^(?!{-).*-}$)` [demo](https://regex101.com/r/JQYeL8/1)? Yes, I know it's probably terrible regexp but I've been always bad at that 🤣 .

> Also, for consistency with `--replace-headers` and `--drop-headers`, it might make sense to add `--add-headers`

Yes, this is something I realized too, #38 will solve this and will be part of the [0.2.0.0](https://github.com/vaclavsvejcar/headroom/milestone/4) release.

**Comment by mheinzel** [comment]
>  I also plan to do some extensive testing before releasing the final 0.2.0.0 as I don't want to have public release that breaks code

Sounds good. Thanks for the effort you are putting into this!

> Well, this is gonna be the tough one to solve...

I can imagine.

As I mentioned, for me/us the current behaviour seems okay, as I would want to format these things as single-line anyways (splitting into one `OPTIONS_GHC` per flag if it gets too long).

Not sure exactly how the regex is supposed to behave (and it's not going to be part of the default config anyways I guess), but would just looking for  `#-}` also do the trick?

Thinking out loud, while I'm happy with the results I can achieve with some simple config for `put-before` and `put-after`, I'm not sure I'm 100% happy with the semantics. I checked the source now, but before that I was not sure how exactly they behave (e.g. when interleaving matches and non-matches, or what happens when there is a `put-before` match before a `put-after` match). I'll think about it a bit, maybe I'll have an idea.

**Comment by vaclavsvejcar** [comment]
> Sounds good. Thanks for the effort you are putting into this!

Thank you for help with testing and for ideas for improvements.

> ...but would just looking for `#-`} also do the trick?

Good point, It indeed should.

> Thinking out loud, ...

Sure this is something that would need be carefully explained in documentation for the final release, but if it looks confusing, it's worth improving. The current solution is the best one I came up with so far, but agree that it's not perfect yet, and definitely is not bulletproof. For exaple - there is still chance that `put-after`/`put-before` will match some pattern inside commented out piece of code, which is obviously wrong and rendering header at such position will likely break the code. I'll probably create issues for all these things and will see if I can come up with some nice solution. Also as my view might be biased (because I'm the one who wrote the code) and there might be things that are obvious for me but confusing for end user, I'll be grateful for any comments and/or ideas for improvement. Let me know if you'll figure out anything 👍

**Comment by vaclavsvejcar** [comment]
@chshersh In order to close this issue as resolved, can you please confirm that the issue you had are solved when trying the current version from `master`? Guess you'd also need to configure Headroom to put the license headers after the pragmas:

```yaml
license-headers:
  haskell:
    put-after: ["^{-#"]
    margin-before: 1 # you might want this
    margin-after: 1  # and this one too
```

Surely I'll document all these configuration options before publishing the final `0.2.0.0` release.

**Comment by chshersh** [comment]
@vaclavsvejcar I'll try running `headroom` on Summoner again to see how it works.

**Comment by vaclavsvejcar** [comment]
@chshersh I'll close this issue for now as the version `0.2.0.0` has been already released. Feel free to reopen or open new issue if you'll still have issues with this. Also please let me know if you plan to use `headroom` for the _Summoner` :-)

**Comment by chshersh** [comment]
@vaclavsvejcar Thanks for your effort! I'm definitely going to try the newest `headroom` version again in the nearest future. Just a bit busy with other plans :slightly_smiling_face:
